lib/Reservation.rb

```diff
@@ -0,0 +1,99 @@
+require 'Date'
+
+module Hotel
+  class Reservation
+
+    attr_reader :block_id, :start_date, :end_date, :room_num
+
+    @@reservations = []
+
+    def initialize(block_id = nil, start_date = Date.today, end_date = Date.today + 1, room_num = 0)
+      check_input(block_id, start_date, end_date, room_num)
+      @block_id = block_id
+      set_reservation_dates(block_id, start_date, end_date)
+      room_num = Reservation.sample_available_rooms(start_date, end_date, block_id).sample if room_num == 0
+      prevent_booking_blocked(room_num, block_id)
+      prevent_doublebooking(start_date, end_date, room_num)
+      @room_num = room_num
+      collect_instance
+    end
+
+    def self.all(date = nil)
+      return @@reservations if date == nil
+      return @@reservations.select { |reservation| reservation.start_date <= date && reservation.end_date >= date }
```

Comment:
I really like the way you have this `.all` method written.  The `select` is a really nice touch.